src/Microsoft.AspNetCore.Mvc.Core/Internal/MiddlewarePipelineService.cs

```diff
@@ -0,0 +1,50 @@
+﻿using System;
+using System.Collections.Concurrent;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Mvc.Internal
+{
+    /// <summary>
+    /// Builds a middleware pipeline after receiving the pipeline from a pipeline provider
+    /// </summary>
+    public class MiddlewarePipelineBuilderService
+    {
+        private readonly ConcurrentDictionary<Type, RequestDelegate> _pipelinesCache
+            = new ConcurrentDictionary<Type, RequestDelegate>();
+
+        public IApplicationBuilder ApplicationBuilder { get; set; }
+
+        public RequestDelegate GetPipeline(Type type)
+        {
+            // Build the pipeline only once. This is similar to how middlewares are used where they are constructed
+            // only once.
+
+            RequestDelegate requestDelegate;
+            if (!_pipelinesCache.TryGetValue(type, out requestDelegate))
+            {
+                var pipelineProvider = (IMiddlewarePipelineProvider)Activator.CreateInstance(type);
```

Comment:
Wut? Dependency injection all the things. Register these things in the container and just call GetService on it.